third_party/maya/lib/usdMaya/MayaParticleWriter.cpp

```diff
@@ -0,0 +1,362 @@
+        for (auto i = decltype(count){0}; i < count; ++i) {
+            ret->operator[](i) = static_cast<T>(a[i]);
+        }
+        return _sharedVtArray<T>(ret);
+    }
+
+    template <typename T>
+    using _strVecPair = std::pair<TfToken, _sharedVtArray<T>>;
+
+    template <typename T>
+    using _strVecPairVec = std::vector<_strVecPair<T>>;
+
+    template <typename T>
+    size_t _minCount(const _strVecPairVec<T>& a) {
+        auto mn = std::numeric_limits<size_t>::max();
+        for (const auto& v : a) {
+            mn = std::min(mn, v.second->size());
+        }
+
+        return mn;
+    }
+
+    template <typename T>
+    void _resizeVectors(_strVecPairVec<T>& a, size_t size) {
+        for (const auto& v : a) {
+            v.second->resize(size);
+        }
+    }
+
+    template <typename T>
+    inline void _addAttr(UsdGeomPoints& points, const TfToken& name, const SdfValueTypeName& typeName,
+                         const VtArray<T>& a, const UsdTimeCode& usdTime) {
+        auto attr = points.GetPrim().CreateAttribute(name, typeName, false, SdfVariabilityVarying);
+        attr.Set(a, usdTime);
+    }
+
+    const TfToken _rgbName("rgb");
+    const TfToken _emissionName("emission");
+    const TfToken _opacityName("opacity");
+    const TfToken _lifespanName("lifespan");
+    const TfToken _massName("mass");
+
+    template <typename T>
+    void _addAttrVec(UsdGeomPoints& points, const SdfValueTypeName& typeName, const _strVecPairVec<T>& a,
+                     const UsdTimeCode& usdTime) {
+        for (const auto& v : a) {
+            _addAttr(points, v.first, typeName, *v.second, usdTime);
+        }
+    }
+
+    // The logic of filtering the user attributes is based on partio4Maya/PartioExport.
+    // https://github.com/redpawfx/partio/blob/redpawfx-rez/contrib/partio4Maya/scripts/partioExportGui.mel
+    // We either don't want these or already export them using one of the builtin functions.
+    const std::set<std::string> _supressedAttrs = {
+        "emitterDataPosition", "emitterDataVelocity", "fieldDataMass",
+        "fieldDataPosition", "fieldDataVelocity", "inputGeometryPoints",
+        "lastCachedPosition", "lastPosition", "lastVelocity",
+        "lastWorldPosition", "lastWorldVelocity", "worldVelocityInObjectSpace",
+        "position", "velocity", "acceleration", "rgb", "rgbPP", "incandescencePP",
+        "radius", "radiusPP", "age", "opacity", "opacityPP", "lifespan", "lifespanPP",
+        "id", "particleId", "mass",
+    };
+
+    // All the initial state attributes end with 0
+    bool _isInitialAttribute(const std::string& attrName) {
+        return attrName.back() == '0';
+    }
+
+    bool _isCachedAttribute(const std::string& attrName) {
+        constexpr auto _cached = "cached";
+        constexpr auto _Cache = "Cache";
+        const auto attrNameSize = attrName.size();
+        if (attrNameSize > 5 && attrName.substr(attrNameSize - 6) == _Cache) {
```

Comment:
I think this could be:
``return TfStringEndsWith(attrName, _Cache) || TfStringStartsWith(attrName, _cached);``

(need to include "pxr/base/tf/stringUtils.h")